svcat: Allow users to delete multiple bindings by using multiple --name flags

luksa opened the pull request.

carolynvs commented on cmd/svcat/binding/unbind_cmd.go:
```diff
@@ -84,28 +84,31 @@ func (c *unbindCmd) Run() error {
-	if err != nil {
-		return err
-	}
+	for _, bindingName := range c.bindingNames {
```
There is existing logic to handle deleting multiple bindings, so we don't need to duplicate it here. Checkout https://github.com/kubernetes-incubator/service-catalog/blob/cd1c2fd0fe61f20196300fe004aedba9dc1eaa73/pkg/svcat/service-catalog/binding.go#L106 and I bet we can extract out that logic so that we can reuse it between unbind instance and deleting multiple bindings.

FYI, the goal is to not have any interesting non-cli logic in the cmd/svcat package. Anything that may be worth resusing should go into pkg/svcat. So whenever you are looking to add functionality to svcat, that's the place to put it.

luksa commented:
@carolynvs Yup, I saw the `unbind INSTANCENAME`. This feature was actually requested by our QE team. I thought it made sense, so I implemented it quickly - without really around too much. 

Anyway, I've refactored the code so it now uses parts of the existing code.

One thing I noticed is that the `unbind INSTANCENAME` command only printed the names of the bindings it deleted if you used the `--wait` switch. I've fixed this, but it's not ideal. I didn't want to completely redo the code that's in `waitIfRequested()` now. 

Also, there was no existing test for unbind (apart from one testing unbind instance). I've now added one that uses multiple bindings. PTAL

carolynvs commented:
> One thing I noticed is that the unbind INSTANCENAME command only printed the names of the bindings it deleted if you used the --wait switch

Yikes! That was a regression that I introduced. Let me fix that today, and add a test for it. I don't want that mixed up in your new feature PR.

carolynvs commented:
Here's a fix for the unbind output: https://github.com/kubernetes-incubator/service-catalog/pull/2049

luksa commented:
Rebased. @carolynvs, please review.

carolynvs commented on cmd/svcat/binding/unbind_cmd.go:
```diff
@@ -55,10 +54,10 @@ func NewUnbindCmd(cxt *command.Context) *cobra.Command {
 	command.AddNamespaceFlags(cmd.Flags(), false)
-	cmd.Flags().StringVar(
-		&unbindCmd.bindingName,
+	cmd.Flags().StringArrayVar(
```
FYI, there is a known bug in cobra around StringArray types. Switch this to StringSliceVar and everything looks the same but avoids the bug.

luksa replied:
Thanks. Fixed.

carolynvs commented on cmd/svcat/binding/unbind_cmd.go:
```diff
@@ -131,16 +109,27 @@ func (c *unbindCmd) unbindInstance() error {
 }
 
+func (c *unbindCmd) getBindingsToDelete() []types.NamespacedName {
+	bindings := []types.NamespacedName{}
```
TIL about NamespacedName 👍

carolynvs commented on cmd/svcat/binding/unbind_cmd.go:
```diff
@@ -131,16 +109,27 @@ func (c *unbindCmd) unbindInstance() error {
 // waitForBindingDeletes waits for the bindings to be deleted and prints either
 // and error message or the name of the deleted binding.
-func (c *unbindCmd) waitForBindingDeletes(waitMessage string, bindings ...v1beta1.ServiceBinding) bool {
+func (c *unbindCmd) waitForBindingDeletes(waitMessage string, bindings ...types.NamespacedName) bool {
```
This is a nice improvement! 💯

carolynvs commented on cmd/svcat/binding/unbind_cmd.go:
```diff
@@ -131,16 +109,27 @@ func (c *unbindCmd) unbindInstance() error {
 	var hasErrors bool
 
+	// Used to prevent concurrent writes to c.Output
+	var mutex sync.Mutex
```
Good catch! I had forgotten that Output is a raw Writer and not a Logger. I may circle back after this goes in to switch so that it's safe to pass around and use in goroutines.

carolynvs commented on cmd/svcat/binding/unbind_cmd_test.go:
```diff
@@ -178,9 +217,32 @@ func TestUnbindCommand(t *testing.T) {
 	}
 }
+
+func outputMatches(gotOutput string, wantOutput string, allowDifferentLineOrder bool) bool {
```
🤗 This is just what these tests needed!

luksa replied:
Yeah, it's not perfect (because the first line with --wait should always be "waiting for the binding(s) to be deleted..."), but I feel handling this would complicate things too much.

carolynvs commented:
I've kicked the build to workaround the timeout.
